File: engine-content/src/main/kotlin/com/palmastro/content/GuidanceBuilder.kt
package com.palmastro.content

import com.palmastro.contracts.Domains
import com.palmastro.contracts.SemanticPayload
import kotlinx.serialization.Serializable
import kotlin.math.abs

/**
 * One "lean into" or "be mindful of" guidance card. [signalId] is the ruleset
 * signal that backed the card, or null when it came from the per-domain
 * bucket generic fallback.
 */
@Serializable
data class GuidanceItem(
    val domain: String,
    val signalId: String?,
    val title: String,
    val body: String,
    val action: String,
)

/**
 * The composed "Understand your reading" guidance layer for one monthly
 * result: a grade-keyed month theme, up to three strengths to lean into,
 * two-to-three gentle mindful pointers, and one weekly focus line per domain.
 */
@Serializable
data class Guidance(
    val monthTheme: String,
    val strengths: List<GuidanceItem>,
    val mindful: List<GuidanceItem>,
    val weekPlan: List<String>,
)

/**
 * Pure, deterministic guidance composer (PRD §11-§13, §30-§32). All copy
 * comes from [ContentTemplates.guidance]; derivation:
 *
 * - strengths: positive explainability contributions sorted descending
 *   (ties: domain order, then signalId) -> top [MAX_STRENGTHS] with distinct
 *   domains; short lists are backfilled with the domain bucket generic for
 *   the highest-scoring uncovered domains.
 * - mindful: negative contributions sorted by magnitude (same tie-breaks),
 *   deduplicated by signalId -> up to [MAX_MINDFUL]; backfilled to at least
 *   [MIN_MINDFUL] with bucket generics, preferring domains not already used
 *   as strengths, lowest score first.
 * - weekPlan: one focus line per domain in [Domains.ALL] order, choosing the
 *   "high"/"low" variant by `scoreCard.totalScore >= 65`.
 *
 * Tone is positivity-first and never fear-based: mindful items are gentle
 * attention-pointers with a concrete micro-action, never warnings of doom.
 */
class GuidanceBuilder(
    private val templates: ContentTemplates = ContentTemplates.default(),
) {

    private data class Candidate(
        val domain: String,
        val domainIndex: Int,
        val signalId: String,
        val contribution: Double,
    )

    fun build(
        payloads: Map<String, SemanticPayload>,
        overallGrade: String,
        language: String,
    ): Guidance {
        val lang = templates.resolveLanguage(language)
        val candidates = collectCandidates(payloads)
        val strengths = buildStrengths(candidates, payloads, lang)
        val mindful = buildMindful(candidates, payloads, strengths, lang)
        return Guidance(
            monthTheme = monthTheme(overallGrade, lang),
            strengths = strengths,
            mindful = mindful,
            weekPlan = weekPlan(payloads, lang),
        )
    }

    private fun collectCandidates(payloads: Map<String, SemanticPayload>): List<Candidate> =
        Domains.ALL.flatMapIndexed { index, domain ->
            payloads[domain]?.explainability.orEmpty()
                .filter { it.mapping.contains(domain) }
                .map { Candidate(domain, index, it.signalId, it.contribution) }
        }

    private fun buildStrengths(
        candidates: List<Candidate>,
        payloads: Map<String, SemanticPayload>,
        lang: String,
    ): List<GuidanceItem> {
        val positives = candidates.filter { it.contribution > 0 }.sortedWith(
            compareByDescending<Candidate> { it.contribution }
                .thenBy { it.domainIndex }
                .thenBy { it.signalId },
        )
        val items = pickSignalItems(positives, MAX_STRENGTHS, lang, keyFor = { it.domain }) {
            templates.guidance.signals[it.signalId]?.leanInto
        }
        val usedDomains = items.map { it.domain }.toSet()
        val backfillOrder = Domains.ALL.withIndex()
            .filter { (_, domain) -> domain !in usedDomains && domain in payloads }
            .sortedWith(
                compareByDescending<IndexedValue<String>> { score(payloads, it.value) }
                    .thenBy { it.index },
            )
            .map { it.value }
        return backfill(items, backfillOrder, MAX_STRENGTHS, lang) { domain ->
            bucketCopy(templates.guidance.domains[domain]?.strengths, score(payloads, domain))
        }
    }

    private fun buildMindful(
        candidates: List<Candidate>,
        payloads: Map<String, SemanticPayload>,
        strengths: List<GuidanceItem>,
        lang: String,
    ): List<GuidanceItem> {
        val negatives = candidates.filter { it.contribution < 0 }.sortedWith(
            compareByDescending<Candidate> { abs(it.contribution) }
                .thenBy { it.domainIndex }
                .thenBy { it.signalId },
        )
        val items = pickSignalItems(negatives, MAX_MINDFUL, lang, keyFor = { it.signalId }) {
            templates.guidance.signals[it.signalId]?.mindfulOf
        }
        val usedDomains = items.map { it.domain }.toSet()
        val strengthDomains = strengths.map { it.domain }.toSet()
        val backfillOrder = Domains.ALL.withIndex()
            .filter { (_, domain) -> domain !in usedDomains && domain in payloads }
            .sortedWith(
                compareBy<IndexedValue<String>> { it.value in strengthDomains }
                    .thenBy { score(payloads, it.value) }
                    .thenBy { it.index },
            )
            .map { it.value }
        return backfill(items, backfillOrder, MIN_MINDFUL, lang) { domain ->
            bucketCopy(templates.guidance.domains[domain]?.mindful, score(payloads, domain))
        }
    }

    /**
     * Walks [sorted] in order, keeping at most [cap] items; a candidate is
     * skipped when its [keyFor] key was already used or [copyFor] has no copy
     * for it. Keys are marked used only on a successful pick, matching the
     * derivation contract mirrored on iOS.
     */
    private fun pickSignalItems(
        sorted: List<Candidate>,
        cap: Int,
        lang: String,
        keyFor: (Candidate) -> String,
        copyFor: (Candidate) -> GuidanceCopy?,
    ): List<GuidanceItem> {
        val used = mutableSetOf<String>()
        val items = mutableListOf<GuidanceItem>()
        sorted.asSequence()
            .takeWhile { items.size < cap }
            .forEach { candidate ->
                val key = keyFor(candidate)
                if (key !in used) {
                    copyFor(candidate)?.let { copy ->
                        items += item(candidate.domain, candidate.signalId, copy, lang)
                        used += key
                    }
                }
            }
        return items
    }

    /** Appends domain-generic items from [order] until [picked] reaches [target]. */
    private fun backfill(
        picked: List<GuidanceItem>,
        order: List<String>,
        target: Int,
        lang: String,
        copyFor: (String) -> GuidanceCopy?,
    ): List<GuidanceItem> {
        val items = picked.toMutableList()
        order.asSequence()
            .takeWhile { items.size < target }
            .forEach { domain ->
                copyFor(domain)?.let { items += item(domain, null, it, lang) }
            }
        return items
    }

    private fun weekPlan(payloads: Map<String, SemanticPayload>, lang: String): List<String> =
        Domains.ALL.mapNotNull { domain ->
            val payload = payloads[domain] ?: return@mapNotNull null
            val key = if (payload.scoreCard.totalScore >= HIGH_THRESHOLD) "high" else "low"
            templates.guidance.domains[domain]?.monthPlan?.get(key)
                ?.let { templates.localized(it, lang) }
                ?.takeIf { it.isNotBlank() }
                ?.replace(DOMAIN_PLACEHOLDER, displayName(domain, lang))
        }

    private fun monthTheme(overallGrade: String, lang: String): String {
        val theme = templates.guidance.monthTheme[overallGrade]
            ?: templates.guidance.monthTheme[DEFAULT_GRADE]
            ?: emptyMap()
        return templates.localized(theme, lang)
    }

    private fun item(domain: String, signalId: String?, copy: GuidanceCopy, lang: String): GuidanceItem {
        val name = displayName(domain, lang)
        fun resolve(text: LocalizedText): String =
            templates.localized(text, lang).replace(DOMAIN_PLACEHOLDER, name)
        return GuidanceItem(
            domain = domain,
            signalId = signalId,
            title = resolve(copy.title),
            body = resolve(copy.body),
            action = resolve(copy.action),
        )
    }

    private fun displayName(domain: String, lang: String): String =
        templates.domains[domain]?.displayName
            ?.let { templates.localized(it, lang) }
            ?.takeIf { it.isNotBlank() }
            ?: domain

    private fun score(payloads: Map<String, SemanticPayload>, domain: String): Int =
        payloads[domain]?.scoreCard?.totalScore ?: DEFAULT_SCORE

    /** First bucket (in template JSON order) whose range contains [score]. */
    private fun bucketCopy(field: Map<String, GuidanceCopy>?, score: Int): GuidanceCopy? =
        field?.entries
            ?.firstOrNull { (bucketId, _) -> templates.buckets[bucketId]?.contains(score) == true }
            ?.value

    companion object {
        private const val MAX_STRENGTHS = 3
        private const val MAX_MINDFUL = 3
        private const val MIN_MINDFUL = 2
        private const val HIGH_THRESHOLD = 65
        private const val DEFAULT_SCORE = 50
        private const val DEFAULT_GRADE = "Stable"
        private const val DOMAIN_PLACEHOLDER = "{domain}"
    }
}

File: ios/PalmAstroKit/Sources/ContentEngine/GuidanceBuilder.swift
import Foundation
import CoreContracts

// Mirrors engine-content/src/main/kotlin/com/palmastro/content/GuidanceBuilder.kt
// exactly ("Understand your reading" guidance layer, templates v2.1.0).
// Parity fixtures in ios/shared-fixtures/guidance assert both implementations
// agree, including item selection order.

/// One "lean into" or "be mindful of" guidance card. `signalId` is the ruleset
/// signal that backed the card, or nil when it came from the per-domain bucket
/// generic fallback.
public struct GuidanceItem: Codable, Equatable, Sendable {
    public let domain: String
    public let signalId: String?
    public let title: String
    public let body: String
    public let action: String

    public init(domain: String, signalId: String?, title: String, body: String, action: String) {
        self.domain = domain
        self.signalId = signalId
        self.title = title
        self.body = body
        self.action = action
    }

    public init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        domain = try c.decode(String.self, forKey: .domain)
        signalId = try c.decodeIfPresent(String.self, forKey: .signalId)
        title = try c.decodeIfPresent(String.self, forKey: .title) ?? ""
        body = try c.decodeIfPresent(String.self, forKey: .body) ?? ""
        action = try c.decodeIfPresent(String.self, forKey: .action) ?? ""
    }
}

/// The composed "Understand your reading" guidance layer for one monthly
/// result: a grade-keyed month theme, up to three strengths to lean into,
/// two-to-three gentle mindful pointers, and one weekly focus line per domain.
public struct Guidance: Codable, Equatable, Sendable {
    public let monthTheme: String
    public let strengths: [GuidanceItem]
    public let mindful: [GuidanceItem]
    public let weekPlan: [String]

    public init(monthTheme: String, strengths: [GuidanceItem], mindful: [GuidanceItem], weekPlan: [String]) {
        self.monthTheme = monthTheme
        self.strengths = strengths
        self.mindful = mindful
        self.weekPlan = weekPlan
    }

    public init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        monthTheme = try c.decodeIfPresent(String.self, forKey: .monthTheme) ?? ""
        strengths = try c.decodeIfPresent([GuidanceItem].self, forKey: .strengths) ?? []
        mindful = try c.decodeIfPresent([GuidanceItem].self, forKey: .mindful) ?? []
        weekPlan = try c.decodeIfPresent([String].self, forKey: .weekPlan) ?? []
    }

    public var isEmpty: Bool {
        monthTheme.isBlank && strengths.isEmpty && mindful.isEmpty && weekPlan.isEmpty
    }
}

/// Pure, deterministic guidance composer (PRD §11-§13, §30-§32). All copy
/// comes from `ContentTemplates.guidance`; derivation:
///
/// - strengths: positive explainability contributions sorted descending
///   (ties: domain order, then signalId) -> top `maxStrengths` with distinct
///   domains; short lists are backfilled with the domain bucket generic for
///   the highest-scoring uncovered domains.
/// - mindful: negative contributions sorted by magnitude (same tie-breaks),
///   deduplicated by signalId -> up to `maxMindful`; backfilled to at least
///   `minMindful` with bucket generics, preferring domains not already used
///   as strengths, lowest score first.
/// - weekPlan: one focus line per domain in `Domains.all` order, choosing the
///   "high"/"low" variant by `scoreCard.totalScore >= 65`.
///
/// Tone is positivity-first and never fear-based: mindful items are gentle
/// attention-pointers with a concrete micro-action, never warnings of doom.
///
/// Kotlin's sorts are stable; Swift's is not guaranteed to be, so every
/// comparator ends in an explicit total tie-break (domain index, then
/// signalId) — the same lesson as the scoring engine's explainability sort.
public struct GuidanceBuilder {

    private static let maxStrengths = 3
    private static let maxMindful = 3
    private static let minMindful = 2
    private static let highThreshold = 65
    private static let defaultScore = 50
    private static let defaultGrade = "Stable"
    private static let domainPlaceholder = "{domain}"

    private let templates: ContentTemplates

    public init(templates: ContentTemplates) {
        self.templates = templates
    }

    /// Convenience initializer loading the bundled default templates.
    public init() throws {
        self.init(templates: try ContentTemplates.loadDefault())
    }

    public func build(
        payloads: [String: SemanticPayload],
        overallGrade: String,
        language: String
    ) -> Guidance {
        let lang = templates.resolveLanguage(language)
        let candidates = collectCandidates(payloads: payloads)
        let strengths = buildStrengths(candidates: candidates, payloads: payloads, lang: lang)
        let mindful = buildMindful(candidates: candidates, payloads: payloads, strengths: strengths, lang: lang)
        return Guidance(
            monthTheme: monthTheme(overallGrade: overallGrade, lang: lang),
            strengths: strengths,
            mindful: mindful,
            weekPlan: weekPlan(payloads: payloads, lang: lang)
        )
    }

    // MARK: - Candidates

    private struct Candidate {
        let domain: String
        let domainIndex: Int
        let signalId: String
        let contribution: Double
    }

    private func collectCandidates(payloads: [String: SemanticPayload]) -> [Candidate] {
        Domains.all.enumerated().flatMap { index, domain -> [Candidate] in
            (payloads[domain]?.explainability ?? [])
                .filter { $0.mapping.contains(domain) }
                .map { Candidate(domain: domain, domainIndex: index, signalId: $0.signalId, contribution: $0.contribution) }
        }
    }

    // MARK: - Strengths (lean into)

    private func buildStrengths(
        candidates: [Candidate],
        payloads: [String: SemanticPayload],
        lang: String
    ) -> [GuidanceItem] {
        var items: [GuidanceItem] = []
        var usedDomains = Set<String>()
        let positives = candidates.filter { $0.contribution > 0 }.sorted { a, b in
            if a.contribution != b.contribution { return a.contribution > b.contribution }
            if a.domainIndex != b.domainIndex { return a.domainIndex < b.domainIndex }
            return a.signalId < b.signalId
        }
        for candidate in positives {
            if items.count == Self.maxStrengths { break }
            if usedDomains.contains(candidate.domain) { continue }
            guard let copy = templates.guidance.signals[candidate.signalId]?.leanInto else { continue }
            items.append(item(domain: candidate.domain, signalId: candidate.signalId, copy: copy, lang: lang))
            usedDomains.insert(candidate.domain)
        }
        if items.count < Self.maxStrengths {
            let backfill = Domains.all.enumerated()
                .filter { !usedDomains.contains($0.element) && payloads[$0.element] != nil }
                .sorted { a, b in
                    let scoreA = score(payloads: payloads, domain: a.element)
                    let scoreB = score(payloads: payloads, domain: b.element)
                    if scoreA != scoreB { return scoreA > scoreB }
                    return a.offset < b.offset
                }
            for (_, domain) in backfill {
                if items.count == Self.maxStrengths { break }
                guard let copy = bucketCopy(
                    field: templates.guidance.domains[domain]?.strengths,
                    score: score(payloads: payloads, domain: domain)
                ) else { continue }
                items.append(item(domain: domain, signalId: nil, copy: copy, lang: lang))
                usedDomains.insert(domain)
            }
        }
        return items
    }

    // MARK: - Mindful (be mindful of)

    private func buildMindful(
        candidates: [Candidate],
        payloads: [String: SemanticPayload],
        strengths: [GuidanceItem],
        lang: String
    ) -> [GuidanceItem] {
        var items: [GuidanceItem] = []
        var usedSignals = Set<String>()
        var usedDomains = Set<String>()
        let negatives = candidates.filter { $0.contribution < 0 }.sorted { a, b in
            let magA = abs(a.contribution)
            let magB = abs(b.contribution)
            if magA != magB { return magA > magB }
            if a.domainIndex != b.domainIndex { return a.domainIndex < b.domainIndex }
            return a.signalId < b.signalId
        }
        for candidate in negatives {
            if items.count == Self.maxMindful { break }
            if usedSignals.contains(candidate.signalId) { continue }
            guard let copy = templates.guidance.signals[candidate.signalId]?.mindfulOf else { continue }
            items.append(item(domain: candidate.domain, signalId: candidate.signalId, copy: copy, lang: lang))
            usedSignals.insert(candidate.signalId)
            usedDomains.insert(candidate.domain)
        }
        if items.count < Self.minMindful {
            let strengthDomains = Set(strengths.map(\.domain))
            let backfill = Domains.all.enumerated()
                .filter { !usedDomains.contains($0.element) && payloads[$0.element] != nil }
                .sorted { a, b in
                    // Prefer domains not already used as strengths, then
                    // lowest score, then canonical domain order.
                    let inStrengthsA = strengthDomains.contains(a.element)
                    let inStrengthsB = strengthDomains.contains(b.element)
                    if inStrengthsA != inStrengthsB { return !inStrengthsA }
                    let scoreA = score(payloads: payloads, domain: a.element)
                    let scoreB = score(payloads: payloads, domain: b.element)
                    if scoreA != scoreB { return scoreA < scoreB }
                    return a.offset < b.offset
                }
            for (_, domain) in backfill {
                if items.count >= Self.minMindful { break }
                guard let copy = bucketCopy(
                    field: templates.guidance.domains[domain]?.mindful,
                    score: score(payloads: payloads, domain: domain)
                ) else { continue }
                items.append(item(domain: domain, signalId: nil, copy: copy, lang: lang))
                usedDomains.insert(domain)
            }
        }
        return items
    }

    // MARK: - Week plan / month theme

    private func weekPlan(payloads: [String: SemanticPayload], lang: String) -> [String] {
        Domains.all.compactMap { domain -> String? in
            guard let payload = payloads[domain] else { return nil }
            let key = payload.scoreCard.totalScore >= Self.highThreshold ? "high" : "low"
            guard let text = templates.guidance.domains[domain]?.monthPlan[key] else { return nil }
            let line = templates.localized(text, language: lang)
            guard !line.isBlank else { return nil }
            return line.replacingOccurrences(
                of: Self.domainPlaceholder,
                with: displayName(domain: domain, lang: lang)
            )
        }
    }

    private func monthTheme(overallGrade: String, lang: String) -> String {
        let theme = templates.guidance.monthTheme[overallGrade]
            ?? templates.guidance.monthTheme[Self.defaultGrade]
            ?? [:]
        return templates.localized(theme, language: lang)
    }

    // MARK: - Helpers

    private func item(domain: String, signalId: String?, copy: GuidanceCopy, lang: String) -> GuidanceItem {
        let name = displayName(domain: domain, lang: lang)
        func resolve(_ text: LocalizedText) -> String {
            templates.localized(text, language: lang)
                .replacingOccurrences(of: Self.domainPlaceholder, with: name)
        }
        return GuidanceItem(
            domain: domain,
            signalId: signalId,
            title: resolve(copy.title),
            body: resolve(copy.body),
            action: resolve(copy.action)
        )
    }

    private func displayName(domain: String, lang: String) -> String {
        let name = templates.domains[domain]
            .map { templates.localized($0.displayName, language: lang) } ?? ""
        return name.isBlank ? domain : name
    }

    private func score(payloads: [String: SemanticPayload], domain: String) -> Int {
        payloads[domain]?.scoreCard.totalScore ?? Self.defaultScore
    }

    /// Bucket whose range contains `score`. Kotlin scans entries in template
    /// JSON order; Swift dictionaries are unordered, so the sorted-key scan
    /// from `ContentTemplates.bucketValue` is used — bucket ranges within one
    /// field are disjoint in the canonical data, so at most one entry matches
    /// and both implementations agree.
    private func bucketCopy(field: [String: GuidanceCopy]?, score: Int) -> GuidanceCopy? {
        guard let field else { return nil }
        return templates.bucketValue(field, score: score)
    }
}
